inference/safety/main.py

```diff
@@ -0,0 +1,57 @@
+# A FastAPI server to run the safety pipeline
+
+import asyncio
+
+import fastapi
+import uvicorn
+from blade2blade import Blade2Blade
+from loguru import logger
+from oasst_shared.schemas import inference
+from settings import settings
+
+app = fastapi.FastAPI()
+
+
+@app.middleware("http")
+async def log_exceptions(request: fastapi.Request, call_next):
+    try:
+        response = await call_next(request)
+    except Exception:
+        logger.exception("Exception in request")
+        raise
+    return response
+
+
+pipeline_loaded: bool = False
+pipeline: Blade2Blade
+
+
+@app.on_event("startup")
+async def load_pipeline():
+    global pipeline_loaded, pipeline
+    pipeline = Blade2Blade(settings.safety_model_name)
+    # warmup
+    input = "|prompter|Hey,how are you?|endoftext|"
+    _ = pipeline.predict(input)
+    pipeline_loaded = True
+    logger.info("Pipeline loaded")
+
+
+@app.post("/safety", response_model=inference.SafetyResponse)
+async def safety(request: inference.SafetyRequest):
+    global pipeline_loaded, pipeline
+    while not pipeline_loaded:
+        await asyncio.sleep(0.1)
+    outputs = pipeline.predict(request.inputs)
```

Comment:
how does the parallelism of this look like? does the async make it absolutely single-threaded/event-looped?

Reply:
Hmm. Because of `async` we can handle multiple requests concurrently, but it's single-threaded and `pipeline.predict()` will still block the thread so I guess we aren't really getting any benefits of parallelism

Reply:
I have implemented a workaround for this, let me know what you think. It's not sophisticated but I think it will help